File: day_18.cpp
/*
 * day_18.cpp
 *
 *  Created on: 20 déc. 2019
 *      Author: Florent
 */

#include <cstdlib>
#include <cstdio>
#include <cctype>

#include <utility>
#include <string>
#include <map>
#include <set>
#include <vector>
#include <fstream>

#include <chrono>

typedef std::pair<int, int> Point;

// directions et point suivant selon une direction
enum Direction {
    Up, Down, Left, Right
};
static inline Point direction_next(Direction d, const Point &p) {
    switch (d) {
    case Up:
        return Point(p.first, p.second - 1);
    case Down:
        return Point(p.first, p.second + 1);
    case Left:
        return Point(p.first - 1, p.second);
    case Right:
        return Point(p.first + 1, p.second);
    }
    return Point(0, 0);
}

class World {
public:
    typedef std::pair<int, std::set<char> > Path; // distance, liste des portes rencontrées
    typedef std::map<char, Path> Paths; // pour chaque point d'intérêt, la liste des chemins vers les autres points d'intérêts

    std::map<Point, char> m_world; // tout le plateau
    std::map<char, Point> m_start_positions; // la ou les (partie 2) positions de départ
    std::map<char, Point> m_keys; // les clés
    std::map<char, Point> m_doors; // les portes
    std::map<char, Paths> m_paths; // tous les chemins qu'on est susceptible de faire avec leurs distances
    std::map<std::string, int> m_rec_cache; // un cache pour les appels à rec_search

    World(const std::string &lines, bool part_2 = false) :
            m_world(), m_start_positions(), m_keys(), m_doors(), m_paths(), m_rec_cache() {
        // parse
        int x = 0;
        int y = 0;
        for (const char c : lines) {
            if (c == '\n') {
                x = 0;
                y += 1;
                continue;
            }
            m_world[Point(x, y)] = c;
            if (c == '@') {
                m_start_positions['@'] = Point(x, y);
            } else if (islower(c)) {
                m_keys[c] = Point(x, y);
            } else if (isupper(c)) {
                m_doors[c] = Point(x, y);
            }
            x += 1;
        }

        // ajustements pour la partie 2, la zone de départ est modifiée et on a 4 robots
        // les points de départ des robots seront @ = $ &
        if (part_2) {
            Point start_pos = m_start_positions['@'];
            m_start_positions['@'] = Point(start_pos.first + 1, start_pos.second + 1);
            m_start_positions['='] = Point(start_pos.first + 1, start_pos.second - 1);
            m_start_positions['$'] = Point(start_pos.first - 1, start_pos.second + 1);
            m_start_positions['&'] = Point(start_pos.first - 1, start_pos.second - 1);
            m_world[m_start_positions['@']] = '@';
            m_world[m_start_positions['=']] = '=';
            m_world[m_start_positions['$']] = '$';
            m_world[m_start_positions['&']] = '&';

            m_world[start_pos] = '#';
            m_world[Point(start_pos.first + 1, start_pos.second + 0)] = '#';
            m_world[Point(start_pos.first - 1, start_pos.second + 0)] = '#';
            m_world[Point(start_pos.first + 0, start_pos.second + 1)] = '#';
            m_world[Point(start_pos.first + 0, start_pos.second - 1)] = '#';
        }

        // précalculer les chemins entre les positions de départ et les clés
        // ainsi qu'entre les différentes clés
        // on stocke la distance et les portes rencontrées (en minuscules)
        for (const auto &s : m_start_positions) {
            m_paths[s.first] = cache_paths(s.second);
        }
        for (const auto &s : m_keys) {
            m_paths[s.first] = cache_paths(s.second);
        }
    }

    int work() {
        // un set avec le ou les points de départ
        std::set<char> start_points;
        for (const auto &p : m_start_positions) {
            start_points.insert(p.first);
        }
        // un set avec toutes les clés à récupérer
        std::set<char> missing_keys;
        for (const auto &p : m_keys) {
            missing_keys.insert(p.first);
        }
        return rec_search(start_points, missing_keys);
    }

private:
    // calcul les chemins vers les autres clés à partir de la position donnée
    Paths cache_paths(const Point &position) {
        // un état pour la recherche
        struct State {
            Point m_position; // position du robot
            std::set<char> m_dependencies; // liste des portes déjà rencontrées

            State(const Point &p, const std::set<char> &d) :
                    m_position(p), m_dependencies(d) {
            }
        };
        constexpr Direction all_dirs[] = { Up, Down, Left, Right };

        std::vector<State> queue; // liste de positions à étudier
        queue.emplace_back(position, std::set<char>());
        std::map<Point, int> visited; // meilleures visites (point, distante)

        Paths paths; // retour

        int distance = 1;

        while (!queue.empty()) {
            decltype(queue) next_queue;
            for (const auto &s : queue) {
                for (const Direction d : all_dirs) {
                    Point next_p = direction_next(d, s.m_position);

                    // a-t-on déjà fait mieux
                    const auto &visited_next_p = visited.find(next_p);
                    if (visited_next_p != visited.end() && visited_next_p->second < distance) {
                        continue;
                    }

                    char v = m_world[next_p];
                    if (v == '#') { // mur
                        continue;
                    }

                    // une clé
                    if (islower(v)) {
                        paths[v] = Path(distance, s.m_dependencies);
                    }

                    // la suite du parcours
                    next_queue.emplace_back(next_p, s.m_dependencies);
                    State &next_s = next_queue.back();

                    // est-ceune porte, si oui mise à jour des dépendances
                    if (isupper(v)) {
                        next_s.m_dependencies.insert(tolower(v));
                    }

                    // enregistrer le passage
                    visited[next_p] = distance;
                }
            }
            queue = next_queue;
            distance += 1;
        }

        return paths;
    }

    int rec_search(const std::set<char> &points_names, const std::set<char> &missing_keys) {
        int best = -1;

        if (missing_keys.size() == 0) {
            return 0;
        }

        // la clé pour le cache d'appel sur rec_search
        // chaîne formée à partir des arguments
        std::string cache_key;
        cache_key.reserve(points_names.size() + missing_keys.size());
        for (char c : points_names) {
            cache_key.push_back(c);
        }
        for (char c : missing_keys) {
            cache_key.push_back(c);
        }
        const auto &cache_test = m_rec_cache.find(cache_key);
        if (cache_test != m_rec_cache.end()) {
            return cache_test->second;
        }

        for (char k : missing_keys) {
            // trouver qui peut chopper la clé k
            for (char bot : points_names) {
                // est-ce que ce bot l'a dans ses chemins possibles
                const auto &path_bot = m_paths[bot];
                const auto &path_search = path_bot.find(k);
                if (path_search == path_bot.end()) {
                    continue;
                }

                // on y va
                const Path &p = path_search->second;
                if (best != -1 and best <= p.first) {
                    // bon c'est pas un champion
                    continue;
                }

                // est-ce qu'on a toutes les clés pour ce chemin
                bool got_all_keys = true;
                for (char missing_key : missing_keys) {
                    if (p.second.count(missing_key)) {
                        got_all_keys = false;
                        break;
                    }
                }
                if (!got_all_keys) {
                    continue;
                }

                // continuer en mettant à jour la position de ce robot
                // et en retirant la clé des clés manquants
                std::set<char> new_search_points = points_names;
                new_search_points.erase(bot);
                new_search_points.insert(k);
                std::set<char> new_missing_keys = missing_keys;
                new_missing_keys.erase(k);
                int r = rec_search(new_search_points, new_missing_keys);
                // progrès ou non ?
                if (best == -1 || (r != -1 && best > p.first + r)) {
                    best = p.first + r;
                }
            }
        }

        m_rec_cache[cache_key] = best;
        return best;
    }

};

int main(int argc, char **argv) {
    const char *opt_filename = "input_18";
    if (argc == 2) {
        opt_filename = argv[1];
    }

    std::string input;
    std::ifstream file(opt_filename);
    if (!file.is_open()) {
        return 1;
    }
    file.seekg(0, std::ios::end);
    input.reserve(file.tellg());
    file.seekg(0, std::ios::beg);

    input.assign((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());

    auto start_p1 = std::chrono::high_resolution_clock::now();
    World world(input, false);
    int r_p1 = world.work();
    auto end_p1 = std::chrono::high_resolution_clock::now();

    printf("part1: %d (%f ms)\n", r_p1, std::chrono::duration<double, std::milli>(end_p1 - start_p1).count());
    printf("taille cache : %zu\n", world.m_rec_cache.size());

    auto start_p2 = std::chrono::high_resolution_clock::now();
    world = World(input, true);
    int r_p2 = world.work();
    auto end_p2 = std::chrono::high_resolution_clock::now();

    printf("part2: %d (%f ms)\n", r_p2, std::chrono::duration<double, std::milli>(end_p2 - start_p2).count());
    printf("taille cache : %zu\n", world.m_rec_cache.size());
}

File: day_18.py
#!/usr/bin/python3
#encoding: UTF-8

import fileinput
import sys
from enum import Enum
import itertools
from collections import namedtuple, OrderedDict

if len(sys.argv) == 1:
    sys.argv += ["input_18"]

class Direction(Enum):
    UP = ord('^')
    DOWN = ord('v')
    LEFT = ord('<')
    RIGHT = ord('>')
    
    def next(self, p):
        if self == Direction.UP:
            return (p[0], p[1] - 1)
        if self == Direction.RIGHT:
            return (p[0] + 1, p[1])
        if self == Direction.DOWN:
            return (p[0], p[1] + 1)
        if self == Direction.LEFT:
            return (p[0] - 1, p[1])

class World(object):
    # si part_2, alors modifications pour 2ème partie
    def __init__(self, lines, part_2=False):
        self.world = {}
        self.start_positions = None
        self.keys = {}
        self.doors = {}
        
        for y, line in enumerate(lines):
            for x, c in enumerate(line.strip()):
                self.world[(x, y)] = c
                if c == '@':
                    self.start_positions = {'@': (x, y)}
                elif c.islower():
                    self.keys[(x, y)] = c
                elif c.isupper():
                    self.doors[(x, y)] = c
        
        
        # modification pour 2ème partie
        if part_2:
            start_pos = self.start_positions['@']
            self.start_positions = {'@': (start_pos[0]+1, start_pos[1]+1),
                                    '=': (start_pos[0]+1, start_pos[1]-1),
                                    '$': (start_pos[0]-1, start_pos[1]+1),
                                    '&': (start_pos[0]-1, start_pos[1]-1)}
            self.world[start_pos] = '#'
            self.world[(start_pos[0]+1, start_pos[1])] = '#'
            self.world[(start_pos[0]-1, start_pos[1])] = '#'
            self.world[(start_pos[0], start_pos[1]+1)] = '#'
            self.world[(start_pos[0], start_pos[1]-1)] = '#'
        
        # nouvelles marques pour les positions de départ : @=$&
        for s in self.start_positions:
            self.world[self.start_positions[s]] = s
        
        
        # précalculer les chemins entre les positions de départ et les clés, 
        # ainsi qu'entre les différentes clés
        # on stocke la distance et les portes rencontrées
        self.paths = {}
        #for s in self.start_positions:
            #self.paths[s] = self.__cache_paths(self.start_positions[s])
        #for key in self.keys:
            #self.paths[self.keys[key]] = self.__cache_paths(key)
        for s in self.start_positions:
            self.__cache_paths_2(s, self.start_positions[s])
        for key in self.keys:
            self.__cache_paths_2(self.keys[key], key)
        
        # un cache pour les nombreux appels à __rec_search
        # pas possible d'utiliser functools.lru_cache car les arguments ne sont pas hashables
        self.__rec_search_call_cache = {}
        
    
    def __cache_paths(self, position):
        if not isinstance(position, type( () )):
            position = tuple(position)
        
        # "structure"
        State = namedtuple("State", ["position", "dependencies"])
        # liste de noeuds à parcourir
        queue = [State(position, set())]
        # noeuds déjà parcouru position -> distance
        visited = {position: 0}
        
        paths = OrderedDict()
        
        distance = 1
        while len(queue) != 0:
            next_queue = []
            for s in queue:
                for d in [Direction.UP, Direction.DOWN, Direction.LEFT, Direction.RIGHT]:
                    next_p = d.next(s.position)
                    if next_p in visited and visited[next_p] < distance: # déjà visité
                        continue
                    v = self.world[d.next(s.position)]
                    if v == "#": # brick wall
                        continue
                    
                    # est-ce une clé
                    if v.islower():
                        paths[v] = (distance, s.dependencies)
                    
                    next_s = State(next_p, s.dependencies.copy())
                    
                    # est-ce une porte
                    if v.isupper():
                        next_s.dependencies.add(v.lower()) # stocker les portes en minuscules pour simplifier les comparaisons
                    
                    # enregistrer le passage
                    visited[next_p] = distance
                    
                    # et continuer
                    next_queue.append(next_s)
            queue = next_queue
            distance += 1
        
        return paths
    
    def __cache_paths_2(self, key_name, position):
        if not isinstance(position, type( () )):
            position = tuple(position)
        
        if not key_name in self.paths:
            self.paths[key_name] = OrderedDict()
        
        # "structure"
        State = namedtuple("State", ["position", "dependencies"])
        # liste de noeuds à parcourir
        queue = [State(position, set())]
        # noeuds déjà parcouru position -> distance
        visited = {position: 0}
                
        distance = 1
        while len(queue) != 0:
            next_queue = []
            for s in queue:
                for d in [Direction.UP, Direction.DOWN, Direction.LEFT, Direction.RIGHT]:
                    next_p = d.next(s.position)
                    if next_p in visited and visited[next_p] < distance: # déjà visité
                        continue
                    v = self.world[d.next(s.position)]
                    if v == "#": # brick wall
                        continue
                    
                    # est-ce une clé
                    if v.islower():
                        self.paths[key_name][v] = (distance, s.dependencies)
                    
                    next_s = State(next_p, s.dependencies.copy())
                    
                    # est-ce une porte
                    if v.isupper():
                        next_s.dependencies.add(v.lower()) # stocker les portes en minuscules pour simplifier les comparaisons
                    
                    # enregistrer le passage
                    visited[next_p] = distance
                    
                    # et continuer
                    next_queue.append(next_s)
            queue = next_queue
            distance += 1
            
    def __rec_search(self, points_names, keyring, missing_keys):
        best = -1
        
        if len(missing_keys) == 0:
            return 0
        
        cache_key = "".join(points_names) + "".join(keyring)
        if cache_key in self.__rec_search_call_cache:
            return self.__rec_search_call_cache[cache_key]
        
        for k in missing_keys:
            # trouver qui peut chopper la clé k
            for bot in points_names:
                # regarder si ce bot peut chopper cette clé
                if not k in self.paths[bot]:
                    continue
            
                # aller chopper k
                distance, dependencies = self.paths[bot][k]
                if best != -1 and best <= distance: # on a déjà mieux
                    continue
                if not keyring.issuperset(dependencies): # on ne peut pas ouvrir les portes pour attendre k
                    continue
                # continuer avec cette clé dans le keyring en remplaçant la position de ce bot par la clé
                r = self.__rec_search(points_names.difference({ bot }).union( { k } ), keyring.union({ k }) , missing_keys - { k })
                # est-ce qu'on progresse
                if best == -1 or (r != -1 and best > distance + r):
                    best = distance + r
        self.__rec_search_call_cache[cache_key] = best
        return best
    
    def work(self):
        res = self.__rec_search(set(self.start_positions.keys()), set(), set(self.keys.values()))
        return res

def test_p1():
    lines="""########################
#f.D.E.e.C.b.A.@.a.B.c.#
######################.#
#d.....................#
########################""".split("\n")
    world = World(lines)
    assert world.work() == 86
    
    lines="""########################
#@..............ac.GI.b#
###d#e#f################
###A#B#C################
###g#h#i################
########################""".split("\n")
    world = World(lines)
    assert world.work() == 81
    
    lines="""#################
#i.G..c...e..H.p#
########.########
#j.A..b...f..D.o#
########@########
#k.E..a...g..B.n#
########.########
#l.F..d...h..C.m#
#################""".split("\n")
    world = World(lines)
    assert world.work() == 136

test_p1()

def p1():
    world = World(fileinput.input(), False)
    print(world.work())
p1()

def test_p2():
    lines="""#######
#a.#Cd#
##...##
##.@.##
##...##
#cB#Ab#
#######""".split("\n")
    world = World(lines, True)
    assert world.work() == 8
    
    lines="""#############
#DcBa.#.GhKl#
#.###...#I###
#e#d#.@.#j#k#
###C#...###J#
#fEbA.#.FgHi#
#############""".split("\n")
    world = World(lines, True)
    assert world.work() == 32
    
test_p2()

def p2():
    world = World(fileinput.input(), True)
    print(world.work())
p2()
